### ISSUE TITLE
Memory leak in the Istio sidecar proxy

### ISSUE THREAD

**Issue by owais** [istio/istio#48028]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

We have been running Istio sidecar proxies that are consistently leaking memory over the span of a few days (as shown in the chart that uses container_memory_working_set_bytes).

<img width="1597" alt="156067608-c1893d08-9f96-463f-9fd6-fcefcbadd910" src="https://github.com/istio/istio/assets/46186/9b6a825b-562c-46e6-aa15-a738b7029868">

We are seeing a slow but constant increase in memory on high traffic pods. Sidecar memory goes from 100-200mb to 1-2GB over the course of a few days and eventually results in the pod getting OOM killed.

Below a chart showing the memory consumption for one of the affected pods. Pod started on 25th Oct with 14% (of 1Gb) memory consumption and today sits at ~60%. In a few more days, it'll result in the pod getting killed once the mem crosses the configured limit of 1GB. We have other cases where this increases much faster. It looks like it the rate of increase is proportional to the traffic.

<img width="980" alt="image" src="https://github.com/istio/istio/assets/46186/43715799-3db7-4757-be63-c4dd282ea01d">

-----

![156458291-62605800-415c-4326-899c-fd34aeb1cad1](https://github.com/istio/istio/assets/46186/e9134a46-ca9a-4924-adf4-8236f0b15c9b)

![156458284-f10c486c-8026-46a5-9be8-bbc77d75f97e](https://github.com/istio/istio/assets/46186/dc650863-295a-43e5-b45f-a3601b8839ae)

### Version

```Text
❯ istioctl version
client version: 1.18.2
control plane version: 1.18.5
data plane version: 1.18.2 (11 proxies), 1.18.3 (18 proxies), 1.18.5 (369 proxies)
```

### Additional Information

I reviewed the output of `istioctl bug-report` and not sure if I can share it publicly. Happy to email/slack it over to a maintainer.

**Comment by hzxuzhonghu** [comment]
Is that because unbounded metrics, which has been reported many times?

**Comment by owais** [comment]
I looked at collected metrics and did not find any obvious high cardinality dimensions. My guess (and hope) is that it is either metrics or orphaned connections. The later is a bit unlikely though as it happens across multiple apps including 3rd party apps like kafka. I'm trying to write a service to allow me to reproduce this on demand so I can test a bunch of variations and see what affects the memory consumption. Once it is ready, will try completely disabling all metrics and see if that helps. 

Appreciate any other tips on how to debug this.

**Comment by owais** [comment]
Update: I disabled all telemetry and within 10 mins memory consumption dropped from 40 - 80% to 10 - 20% on all pods. It has remained flat and has not increased in the last 12 hours or so which is very promising. Will keep monitoring it for a couple of days and then try to close in on which metrics/dimensions are causing the "leak".

**Comment by kyessenov** [comment]
The heap profile doesn't show anything outstanding, probably because it's a slow leak. Are you using the default telemetry configuration? Did you try the metric rotation feature and see if it helps?

**Comment by zirain** [comment]
> @kyessenov I'll try the metric rotation feature. Are you talking about METRIC_ROTATION_INTERVAL & METRIC_GRACEFUL_DELETION_INTERVAL env vars?

yes, but you should use Telemetry API instead of EnvoyFilter.

**Comment by owais** [comment]
Thanks. Is there an example of using this with the telemetry API? All I could find in docs were the env vars which I _think_ should be added to istiod pods?

**Comment by owais** [comment]
Ah, I think you mean using env vars is fine but they won't work if telemetry is configure using EnvoyFilter and will only work if Telemetry API is used?

**Comment by zirain** [comment]
if you're using EnvoyFilterm you can directly setting them. see detail from [here](https://github.com/istio/istio/blob/8f16765a1baa02e651affd083e937bf57017d336/pilot/pkg/model/telemetry.go#L1147)

**Comment by owais** [comment]
Thanks. Does "directly setting them" mean setting the above env vars on istio discovery pods?

**Comment by zirain** [comment]
> Thanks. Is there an example of using this with the telemetry API? All I could find in docs were the env vars which I _think_ should be added to istiod pods.

with istio 1.18+, by default it will not install envoyfilter, but enable default prometheus provider.
you can install istio 1.18 with following iop to enable metric rotation:
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  values:
    pilot:
      env:
        METRIC_ROTATION_INTERVAL: 5m
        METRIC_GRACEFUL_DELETION_INTERVAL: 1m

```

**Comment by owais** [comment]
I don't use istio operator. We use helm and have custom telemetry config to drop some dimensions from some metrics. This custom telemetry config makes the istio helm chart deploy Envoy Filters instead of the Telemetry objects. If I understand the discussion, other issues, code and PRs so far correctly, I need to remove all the custom telemetry config so that the Helm chart uses Telemetry to configure metrics instead of Envoy Filters. I'll look into doing that and deploying custom Telemetry objects to drop dimensions.

What is not clear so far is if I *don't* use the Telemetry API and continue relying on EnvoyFilters, will the metric rotation feature not work at all? Is it only supported by the new Telemetry API?

And if I do manage to switch to the new Telemetry API, will adding the two end vars to `istio-discovery` pods be enough to activate the feature?

**Comment by zirain** [comment]
> What is not clear so far is if I don't use the Telemetry API and continue relying on EnvoyFilters, will the metric rotation feature not work at all? Is it only supported by the new Telemetry API?

Telemetry API is just a way to generate the filter configuration, you can use EnvoyFilter directly, but it's hard to write them correctly, that's why istio provide the first class API(Telemetry API).

**Comment by owais** [comment]
I see. So Telemetry API internally generates EnvoyFilters as well so the env vars on istiod pods should work irrespective of whether Telemetry API is used or not?

Thanks for your patience and help.

**Comment by zirain** [comment]
> I see. So Telemetry API internally generates EnvoyFilters as well so the env vars on istiod pods should work irrespective of whether Telemetry API is used or not?

Telemetry API will directly generates a filter in LDS, will not use EnvoyFilter.

https://github.com/istio/istio/blob/8f16765a1baa02e651affd083e937bf57017d336/pilot/pkg/model/telemetry.go#L1147

**Comment by owais** [comment]
Thanks. Let me take a step back and ask what I really want to figure out. 

Can I use `METRIC_ROTATION_INTERVAL` & `METRIC_GRACEFUL_DELETION_INTERVAL` environment variables if I'm NOT using the Telemetry API? 

Now it seems I must use the Telemetry API in order for these env vars to work. Right or wrong? 

I understand there might be other ways to implementing the same feature with Envoy Filters directly but I'm sepcifically asking about the above two istiod env vars.

**Comment by owais** [comment]
I further analyzed the metric dimensions and although I did not see any obvious ones with potential high cardinality, I tried to remove everything that we did not need in our observability dashboards and things have been pretty good since then. Here are the things I tried.

1. [Nov 23] Drop destination_service and request_host dimensions from sidecars as they had obvious high cardinality issues for some services. Initially it seemed this fixed everything but it was obvious memory was still growing for most services some much slower than before though.
2. [Dec 29] We had not configured all metrics and were missing some (request_messages, response_bytes, etc). I added them to our custom metric configuration as well and dropped all the dimensions we were dropping for other metrics. 
3. [Dec 1] Dropped some more dimensions that were not exactly high cardinality but had a lot of values (destination_port) or we didn't need in our dashboards (connection_security_policy). Only dropped this from some metrics. Did not seem to have a considerable impact as expected.
4. [Dec 3] Identified and dropped more dimensions that we did not strictly need to reduce the number of time series generated for histogram buckets (`response_code`, `response_flags`). Also added `disable_host_header_fallback: true` even though if I understand correctly this should not have had any effect on anything as we were already dropped `request_host` dimensions everywhere. Observed this for about 16 hours and memory was still gradually increasing
5. [Dec 4] Dropped `destination_port` from all metrics and cleaned up some custom metric configuration [1] where we map/transform dimension values. A lot of these dimensions were not in use anymore so dropped the custom config was not an issue. I'm not sure if we even needed it in the first place. Since this patch, we've not seen any increase in memory consumption at all for any service in any environment. I tried to isolate it a couple of times but did not narrow it down to any single config. Note that it takes at least a day or two to notice the leak so trying all combinations is quite time consuming. I'd still like to try and isolate it in the next few weeks as time allows.


Above timeline as a chart:

<img width="1645" alt="overall" src="https://github.com/istio/istio/assets/46186/900b8617-be84-4217-97dc-1ae6698a3cad">



--

Hope this helps others who've run into similar issues and more importantly the Istio team to identify the root cause and perhaps patch and/or document it somewhere as a best practice/FAQ.



----

1: custom metric config example:
<img width="502" alt="image" src="https://github.com/istio/istio/assets/46186/6d268973-b51e-4dae-a5ad-515a32c5fd9c">

**Comment by zirain** [comment]
> [Dec 29] We had not configured all metrics and were missing some (request_messages, response_bytes, etc). I added them to our custom metric configuration as well and dropped all the dimensions we were dropping for other metrics.

is this happened in Nov 29?

**Comment by owais** [comment]
Yes, that was Nov 29. Fixed. 

Thanks

**Comment by owais** [comment]
Note that while dropping high cardinality metrics helped reduce memory usage for some services, it did not fix the "leak". We were able to fix it only after eliminating a bunch of dimensions and config none of which seemed to have boundless cardinality. So it looks like there is still an issue somewhere with telemetry leaking memory in certain situations.

**Comment by kyessenov** [comment]
@owais The stat system is inherently flawed if the tags change over time. Even when counters are 0 the symbols are still stored, so you have to rotate the scope per @zirain suggestion. Do the values of your stats change over time? Because if they don't, then we really have a leak since we should be reusing the symbols.

**Comment by owais** [comment]
@kyessenov do you mean if dimensions/tags see new values over time? That was the case for some services but not all. After I dropped all the dimensions that had a lot of values, most services should've seen all dimension values within a few minutes to maybe couple of hours max. That means memory should've only increased for a couple of hours and then flattened right?

**Comment by hanxiaop** [comment]
@owais Have you tried 1.18.7? https://github.com/istio/istio/releases/tag/1.18.7

**Comment by owais** [comment]
@hanxiaop I'm on vacation. Will try it in a couple of weeks.

**Comment by taitelman** [comment]
I confirm. even in v1.20.4 the telemeterics leak:
here is a typical memory graph of the istio-proxy container in some workload. 
![sidecar-ram-across-hours-leak](https://github.com/istio/istio/assets/8079496/8bb91155-fa9e-4774-aa88-cb19c306601f)

as soon as I tuned off istio telemetrics the memory leak is gone.

**Comment by owais** [comment]
Looks like the patch we tried [here](https://github.com/istio/istio/issues/48028#issuecomment-1850751671) did not fix it for all services or it doesn't work with a newer version of Istio. We are seeing similar leaks in TCP services again after rolling out istio to more clusters. It is possible that it might be totally unrelated to this issue. 

Opening new issue at #52374
